### PR TITLE
Module: Kitsu module

### PR DESCRIPTION
**Pull request by Tilix4**
## Brief description
[Kitsu](https://www.cg-wire.com/fr/kitsu.html) module.

## Description
Use Kitsu as project management module.

## Documentation (add _"type: documentation"_ label)
Doc has been written, you can read it in PR files.

## Testing notes:
1. Fill `Server` url in Project Settings.
2. When launching Tray, `Kitsu Credentials` window will pop. Either way click on `Kitsu Connect` button and sign-in.
3. Run `.poetry/bin/poetry run python start.py module kitsu sync-service -l me@domain.ext -p mypass` to start sync OP with Kitsu, events listening will start right after sync is achieved.
4. Publish a new version of an asset and see the magic happening.

You can also try `.poetry/bin/poetry run python start.py module kitsu push-to-zou -l me@domain.ext -p mypass`, but it cannot handle all use cases and must be considered as a helpful facultative utility.

Listening currently supports most of _project/episode/sequence/shot/asset/task create/update/delete_ events, but not `project:delete`, `episode:delete`, `sequence:delete` because of a bug, which will be resolved by https://github.com/cgwire/zou/pull/427

### CODE REVIEW

**Comment by BigRoy** [comment]
@Tilix4 great to see you are looking into tackling this! 💪 

I didn't see any related issue regarding this so I just wanted to point out this **ancient!** set of tool prototypes I did for `gazu` (Python client for Kitsu) called [`qtazu`](https://github.com/Colorbleed/qtazu). I am not sure how well that code holds up Today but it could be a good reference for maybe kitsu related code if it is still somewhat valid.

**Review by mkolar — CHANGES_REQUESTED** [review]
Awesome to see work being done on Kitsu front. Just an important note for start, that we are trying to get rid of the `default_modules` folder. Could you please move the module one level up?

**Comment by Tilix4** [comment]
> Awesome to see work being done on Kitsu front. Just an important note for start, that we are trying to get rid of the `default_modules` folder. Could you please move the module one level up?
> 
> Actually @iLLiCiTiT just told me that for now this is easier for you, so just keep in mind that it will eventually move please.

Yes I started there because @iLLiCiTiT told me so. I'd be glad to have help to move it one level up :)

**Comment by iLLiCiTiT** [comment]
> I'd be glad to have help to move it one level up :)

When you move it, you have to add `"kitsu"` (name of folder where module is) into `DEFAULT_OPENPYPE_MODULES` variable in `openpype/modules/base.py`.

**Comment by Tilix4** [comment]
@iLLiCiTiT @mkolar  I moved the module up.

I think the module is ready to be tested for basic usage, some things are not working (`project:delete`, `episode:delete`, `sequence:delete`) and are about to be fixed on Kitsu's side: https://github.com/cgwire/zou/pull/427

I also commited the `pyproject.toml` with `gazu`.

I'll write some documentation now but I cannot find any guideline about how to write documentation for OP, did I miss something?

File: openpype/modules/kitsu/__init__.py
Comment (iLLiCiTiT) [inline openpype/modules/kitsu/__init__.py:7]: Please remove `AddonSettingsDef` when settings are in default schemas.

**Comment by mkolar** [comment]
> I'll now write some documentation but I cannot find any guideline about how to write documentation for OP, did I miss something?

There aren't docs about the docs :D. But it is simple enough. We use docusaurus and all it's content is inside https://github.com/pypeclub/OpenPype/tree/develop/website

To add new documents you just add .md pages to /docs folder and add the new entry into sidebars.js.

Docusaurus itself has very good documentation https://docusaurus.io/docs (surprise, surprise :D ) if you want to see more details. 

Other than that, this page shows some example formatings https://github.com/pypeclub/OpenPype/blob/develop/website/docs/admin_docsexamples.md

And if you want to run it locally to test, you'll need to run `yarn install` inside of the website folder and then launch the docs with the provided tool https://github.com/pypeclub/OpenPype/blob/develop/tools/run_documentation.ps1

File: openpype/modules/kitsu/kitsu_widgets.py
Comment (iLLiCiTiT) [inline openpype/modules/kitsu/kitsu_widgets.py:11]: Please don't use local settings for module configurations. Like I've mentioned on discord it is not meant for this kind of data and each modification in local settings UI will remove all data that are stored here. Also local settings are stored in mongo database. Please use `OpenPypeSecureRegistry` or other alternative but not this.

File: openpype/modules/kitsu/kitsu_module.py
Comment (iLLiCiTiT) [inline openpype/modules/kitsu/kitsu_module.py:7]: It is really important that this file contains entrypoint of the module describing the public face but not real implementation. Imports here must not contain any dependencies and functions should not do really complex logic. Imagine that you don't have `gazu` python module installed and expect that any side file of the module can crash but this file still must be possible to import.

For example the whole sync function should be in separated file that is imported only when sync is called.

File: openpype/modules/kitsu/kitsu_module.py
Comment (iLLiCiTiT) [inline openpype/modules/kitsu/kitsu_module.py:77]: This will 100% crash on first run.

File: openpype/modules/kitsu/kitsu_module.py
Comment (iLLiCiTiT) [inline openpype/modules/kitsu/kitsu_module.py:97]: How can you know kitsu login if user did not login yet.

File: openpype/modules/kitsu/kitsu_module.py
Comment (iLLiCiTiT) [inline openpype/modules/kitsu/kitsu_module.py:470]: I don't think this should be cli function but module should handle this on `tray_start` if user is not logged in.

File: openpype/modules/kitsu/kitsu_widgets.py
Comment (Tilix4) [reply openpype/modules/kitsu/kitsu_widgets.py:11]: Sorry if I misunderstood. I'll have a look with `OpenPypeSecureRegistry` but when I tried few days ago I didn't find how to use it.

File: openpype/modules/kitsu/kitsu_module.py
Comment (Tilix4) [reply openpype/modules/kitsu/kitsu_module.py:72]: If I understand you well, the user will have to log-in every time he launches Tray? Doing this was to allow the user to use the `remember` feature for credentials.

File: openpype/modules/kitsu/kitsu_widgets.py
Comment (Tilix4) [reply openpype/modules/kitsu/kitsu_widgets.py]: I was tired for sure, I changed it and it seems to be okay, you'll tell me.

File: openpype/modules/kitsu/kitsu_module.py
Comment (Tilix4) [reply openpype/modules/kitsu/kitsu_module.py]: Fixed

File: openpype/modules/kitsu/utils/sync_service.py
Comment (iLLiCiTiT) [inline openpype/modules/kitsu/utils/sync_service.py:388]: I think you can remove this because it is in command line arguments.

File: openpype/modules/kitsu/kitsu_module.py
Comment (iLLiCiTiT) [inline openpype/modules/kitsu/kitsu_module.py:125]: Remove Python 3 notation
```suggestion
def sync_service(login, password):
```

File: openpype/modules/kitsu/kitsu_module.py
Comment (iLLiCiTiT) [inline openpype/modules/kitsu/kitsu_module.py:108]: Remove Python 3 notation
```suggestion
def push_to_zou(login, password):
```

File: openpype/modules/kitsu/kitsu_module.py
Comment (Tilix4) [reply openpype/modules/kitsu/kitsu_module.py:125]: Sorry I forgot that...

File: openpype/modules/kitsu/plugins/publish/kitsu_plugin.py
Comment (iLLiCiTiT) [inline openpype/modules/kitsu/plugins/publish/kitsu_plugin.py:42]: Please keep Python 2 compatibility in publish plugins.

File: openpype/modules/kitsu/utils/update_op_with_zou.py
Comment (ClementHector) [inline openpype/modules/kitsu/utils/update_op_with_zou.py:259]: It could decrease the data to be synchronized
```suggestion
    all_projects = gazu.project.all_open_projects()
```
And maybe switch closed projects to inactive project in OP.

Comment (Tilix4) [reply openpype/modules/kitsu/utils/update_op_with_zou.py:259]: > And maybe switch closed projects to inactive project in OP.

This would be useful.

**Comment by ClementHector** [comment]
The data put in asset_data["tasks"][TASK_NAME]["zou"] are removed if you make changes to the task in the project manage interface. Do we have to change the place to put the zoo data of the task or modify the behavior of the project manage?

**Comment by Tilix4** [comment]
> The data put in asset_data["tasks"][TASK_NAME]["zou"] are removed if you make changes to the task in the project manage interface. Do we have to change the place to put the zoo data of the task or modify the behavior of the project manage?

That's a great question, maybe @iLLiCiTiT has an answer?

**Comment by mkolar** [comment]
>That's a great question, maybe @iLLiCiTiT has an answer?

@Tilix4 @ClementHector yeah to get around that we'd have to tweak some project manager logic. This two way mapping is actually the reason why we don't currently recommend using ftrack and our project manager at the same time. PM was really added just for micro productions that don't have any third party tool.

**Comment by ClementHector** [comment]
@mkolar @Tilix4 I worked around the problem in the plugins, the collector will look for the data in another way if it no longer exists in `asset_data["tasks"][TASK_NAME]["zou"]` . See https://github.com/Tilix4/OpenPype/pull/3 in `openpype/modules/kitsu/plugins/publish/collect_kitsu_entities.py`

**Comment by Tilix4** [comment]
@iLLiCiTiT regarding the merge of this feature, do you require the publish plugin to be ready before merging, or could we merge the `Kitsu to OP sync` first, and then deal with @ClementHector's work for publishing from OP to Kitsu?

Also, do you think we should keep the kitsu module tagged as `beta` into OP during some times after merge?

File: openpype/modules/kitsu/utils/update_zou_with_op.py
Comment (iLLiCiTiT) [inline openpype/modules/kitsu/utils/update_zou_with_op.py:9]:
```suggestion
from openpype.pipeline import AvalonMongoDB
```

**Comment by iLLiCiTiT** [comment]
> do you require the publish plugin to be ready before merging

No, but I would remove the publish plugins as they're useless now and possibly breaking.

> Also, do you think we should keep the kitsu module tagged as beta into OP during some times after merge?

That's a very good question. I would say that it should be beta for some time.

**Comment by Tilix4** [comment]
> No, but I would remove the publish plugins as they're useless now and possibly breaking.

In fact, @ClementHector work seems to work pretty well, then I propose to merge it into this branch, mark it as ready and merge the kitsu module into your master, as `beta`.

We'll test it much stronger very soon, then I may fix things quickly.

File: openpype/modules/kitsu/utils/update_op_with_zou.py
Comment (iLLiCiTiT) [inline openpype/modules/kitsu/utils/update_op_with_zou.py:42]: You change `AVALON_PROJECT`, which affect how `dbcon` object behaves, then you don't use the changed behavior but return direct access to project's database instead. So the change of `AVALON_PROJECT` is pointless.

NOTE: Ideally, you should not use direct access to database but through object of `AvalonMongoDB`. If you want make sure that project is not changed, then create new object of `AvalonMongoDB`, they share mongo connection.
```
# This is how it was meant to use 'AvalonMongoDB'
dbcon = AvalonMongoDB()

dbcon.Session["AVALON_PROJECT"] = project_name
dbcon.find_one(...)
dbcon.update_one(...)

# This is how you use it
dbcon = AvalonMongoDB()

col = dbcon.database[project_name]
col.find_one(...)
col.update_one(...)
```

Comment (Tilix4) [reply openpype/modules/kitsu/utils/update_op_with_zou.py:42]: You mean it'd be better to not directly use `col` but `Session[]`?

To me it seems cleaner to access directly the database collection instead of using some general session setting no?

Comment (iLLiCiTiT) [reply openpype/modules/kitsu/utils/update_op_with_zou.py:42]: > You mean it'd be better to not directly use col but Session[]?

You should use methods on object of `AvalonMongoDB` instead of direct access of project collection.

> To me it seems cleaner to access directly the database collection instead of using some general session setting no?

It's not general. Each `AvalonMongoDB` object has it's own session. Only general session is  `openpype.pipeline.legacy_io.Session`.

This would work exactly the same but using AvalonMongoDB object:
```
def set_op_project(dbcon, project_id) -> AvalonMongoDB:
    """Set project context.
    :param dbcon: Connection to DB.
    :param project_id: Project zou ID
    """
    project = gazu.project.get_project(project_id)
    project_name = project["name"]
    dbcon.Session["AVALON_PROJECT"] = project_name
    return dbcon
```

And this would be "thread safe" for cases that multiple threads would access the same AvalonMongoDB object for different projects.
```
def get_project_con(project_id) -> Collection:
    project = gazu.project.get_project(project_id)
    project_name = project["name"]
    dbcon = AvalonMongoDB()
    dbcon.Session["AVALON_PROJECT"] = project_name
    return dbcon
```

Comment (Tilix4) [reply openpype/modules/kitsu/utils/update_op_with_zou.py]: I changed everything, could you tell me if it fits your expectations?

File: openpype/modules/kitsu/utils/update_op_with_zou.py
Comment (iLLiCiTiT) [inline openpype/modules/kitsu/utils/update_op_with_zou.py:117]: I would recommend to not get project settings in each loop. The settings probably won't change but resolving project settings can slow the sync down a lot.

File: openpype/modules/kitsu/utils/update_op_with_zou.py
Comment (iLLiCiTiT) [inline openpype/modules/kitsu/utils/update_op_with_zou.py:67]: Looks like this line will add to into asset document's data more then should. For example `frame_in` and `frame_out`...

**Review by iLLiCiTiT — APPROVED** [review]
I would call this a good start point. There can be done some enhancements and fixes, but that can be done in future PRs.

File: openpype/modules/kitsu/utils/update_op_with_zou.py
Comment (Tilix4) [reply openpype/modules/kitsu/utils/update_op_with_zou.py:96]: They have one, but a useless one, they are all called `main`...
The task type name is really the way to get the task's kind.

File: openpype/modules/kitsu/utils/update_op_with_zou.py
Comment (Tilix4) [reply openpype/modules/kitsu/utils/update_op_with_zou.py]: Refactored.

File: openpype/modules/kitsu/utils/update_op_with_zou.py
Comment (Tilix4) [reply openpype/modules/kitsu/utils/update_op_with_zou.py]: Moved out of loop

File: openpype/modules/kitsu/utils/update_op_with_zou.py
Comment (Tilix4) [reply openpype/modules/kitsu/utils/update_op_with_zou.py:81]: poped these useless data

File: openpype/modules/kitsu/plugins/publish/validate_kitsu_intent.py
Comment (iLLiCiTiT) [inline openpype/modules/kitsu/plugins/publish/validate_kitsu_intent.py:19]: It should probably stop processing if intent is not available on context.

File: openpype/modules/kitsu/plugins/publish/validate_kitsu_intent.py
Comment (iLLiCiTiT) [inline openpype/modules/kitsu/plugins/publish/validate_kitsu_intent.py:20]: Intent is meant as "Intent of publishing" not status of task.

Comment (Tilix4) [reply openpype/modules/kitsu/plugins/publish/validate_kitsu_intent.py:20]: I don't get your remark.

File: openpype/modules/kitsu/plugins/publish/validate_kitsu_intent.py
Comment (Tilix4) [reply openpype/modules/kitsu/plugins/publish/validate_kitsu_intent.py:20]: Done.

File: openpype/modules/kitsu/plugins/publish/validate_kitsu_intent.py
Comment (Tilix4) [reply openpype/modules/kitsu/plugins/publish/validate_kitsu_intent.py:21]: You mean, when intent is not `None`, we must avoid updating the status in Kitsu, and when it's `None`, change the status to `Waiting for Approval`?

Comment (Tilix4) [reply openpype/modules/kitsu/plugins/publish/validate_kitsu_intent.py:21]: In our case, this intent code seems out of topic, and status creation must be moved to `IntegrateKitsuNote`.

Comment (Tilix4) [reply openpype/modules/kitsu/plugins/publish/validate_kitsu_intent.py]: I've made the relevant changes

Comment (mkolar) [reply openpype/modules/kitsu/plugins/publish/validate_kitsu_intent.py]: > In our case, this intent code seems out of topic, must be deleted, and status change must be moved to IntegrateKitsuNote.

That's correct. `Intent` is technically a unified comment thing that is requested by certain vendors and studios. For instance when artist publishes first version of a comp it might have status `for review` but `intent` might just be `tech validation`.... later on v03 would  still go to `for review` with intent `wip` , and at the end v06 might be `for review` and intent `for final`... or similar. intents are of course arbitrary in each studio, but most importantly they are static data on a version and don't change with status updates

File: openpype/modules/kitsu/plugins/publish/integrate_kitsu_review.py
Comment (iLLiCiTiT) [inline openpype/modules/kitsu/plugins/publish/integrate_kitsu_review.py:33]: Unnecessary double iteration
```suggestion
        for representation in instance.data.get("representations", []):
            if "review" not in r.get("tags", []):
                continue
```

File: openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py
Comment (iLLiCiTiT) [inline openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:15]: So when intent is not available then note is not integrated -> and review is not integrated?

Comment (Tilix4) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:15]: Hum, actually this is the opposite. If the intent has no value, it's supposed to be a normal publish. If an intent is set (WIP, test), then the Production tracker is not updated but OP is.

Comment (iLLiCiTiT) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:15]: I don't think that intent should affect if note or review is integrated at all. Intent can be `"For review"`, `"Final"` or anything else, dynamically created by studio.

Comment (Tilix4) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:15]: You mean, it shouldn't affect note or review by default, it's up to the studio to decide to override or not this behavior?

Comment (iLLiCiTiT) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:15]: Yes, intent is just an information from artist that should be reflected somewhere, like a "comment" with predefined messages. Should not be used as definition of publish flow, at least not by default.

NOTE: Intent is OpenPype specific thing which is not in pyblish-[base](https://github.com/pyblish/pyblish-base)/[lite](https://github.com/pyblish/pyblish-lite)/[qml](https://github.com/pyblish/pyblish-qml).

Comment (BigRoy) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:15]: > If the intent has no value, it's supposed to be a normal publish. If an intent is set (WIP, test), then the Production tracker is not updated but OP is.

If really needed then this sounds more like a Admin Settings tweak like a "filter" for when to integrate to Kitsu. For example:
- Settings for "Exclude Intents for Integration to Kitsu" could be set to `{"WIP", "test"}`
- The integrator would just check against that `if intent in exclude_intents: return`.

Comment (Tilix4) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py:15]: > Should not be used as definition of publish flow, at least not by default.

Okay, will do it this way.

> If really needed then this sounds more like a Admin Settings tweak like a "filter" for when to integrate to Kitsu.
I don't need it currently, with @ClementHector we thought we should do it this way. Then we'll wait for feedback.

Comment (ClementHector) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py]: Publish comments can be integrated into kitsu/ftrach... with the review, right?

Comment (iLLiCiTiT) [reply openpype/modules/kitsu/plugins/publish/integrate_kitsu_note.py]: > Settings for "Exclude Intents for Integration to Kitsu" could be set to

I would say that this integrator plugin should not look for intent at all, but should have mechanism to skip if "some value" is set (or not set).

Like, what if I don't care about intent at all and just want integrate everything, all the time. With this logic I also have to care about intent even if it should not matter.

**Comment by ClementHector** [comment]
@mkolar Hello, Is there something to do to accept this request?

**Review by mkolar — APPROVED** [review]

**Comment by mkolar** [comment]
I'd say this is ready to go!

**Comment by mkolar** [comment]
Kitsu meet OpenPype, OpenPype meet Kitsu. Play nice together and don't make much mess......

**Comment by mkolar** [comment]
@all-contributors please add @Tilix4 for code and documentation

**Pull request merged**
